### PR TITLE
Feature: add value object for `Policy` for record-keeping

### LINKED ISSUE #1251: DataManagementAPI - ContractNegotiationApiController : Initiate ContractNegotiation by PolicyId not working
# Bug Report

## Describe the Bug

At the time of writing contract negotiation can be initiated by sending a `NegotiationInitiateRequestDto` to the data management API. This object contains a `ContractOfferDescription` which looks something like this

```java
public class ContractOfferDescription {
    @NotNull
    private final String offerId;
    @NotNull
    private final String assetId;
    private final String policyId;
    private final Policy policy;
}
```

In this class ether the `policy` or the `policyId` must be set for the description to be valid. In case the call is done using the `policyId` the connector is logging the following message (in a very high frequency):

```bash
[2022-05-03 14:05:19] [FINE   ] [Consumer] Failed to send contract offer with id 1:foo. ContractNegotiation 1a793260-aefc-47b8-ae22-408fd24dab5d stays in state INITIAL.
java.lang.NullPointerException: Cannot invoke "org.eclipse.dataspaceconnector.policy.model.Policy.getPermissions()" because "policy" is null
	at org.eclipse.dataspaceconnector.ids.transform.ContractOfferToIdsContractOfferTransformer.transform(ContractOfferToIdsContractOfferTransformer.java:63)
	at org.eclipse.dataspaceconnector.ids.transform.ContractOfferToIdsContractOfferTransformer.transform(ContractOfferToIdsContractOfferTransformer.java:36)
	at org.eclipse.dataspaceconnector.ids.core.transform.IdsTransformerRegistryImpl.transform(IdsTransformerRegistryImpl.java:58)
	at org.eclipse.dataspaceconnector.ids.core.transform.IdsTransformerRegistryImpl.transform(IdsTransformerRegistryImpl.java:46)
	at org.eclipse.dataspaceconnector.ids.api.multipart.dispatcher.sender.MultipartContractOfferSender.createContractRequest(MultipartContractOfferSender.java:131)
	at org.eclipse.dataspaceconnector.ids.api.multipart.dispatcher.sender.MultipartContractOfferSender.buildMessagePayload(MultipartContractOfferSender.java:110)
	at org.eclipse.dataspaceconnector.ids.api.multipart.dispatcher.sender.MultipartContractOfferSender.buildMessagePayload(MultipartContractOfferSender.java:44)
	at org.eclipse.dataspaceconnector.ids.api.multipart.dispatcher.sender.IdsMultipartSender.send(IdsMultipartSender.java:150)
	at org.eclipse.dataspaceconnector.ids.api.multipart.dispatcher.sender.MultipartContractOfferSender.send(MultipartContractOfferSender.java:44)
	at org.eclipse.dataspaceconnector.ids.core.message.IdsRemoteMessageDispatcher.send(IdsRemoteMessageDispatcher.java:46)
	at org.eclipse.dataspaceconnector.core.base.RemoteMessageDispatcherRegistryImpl.send(RemoteMessageDispatcherRegistryImpl.java:49)
	at org.eclipse.dataspaceconnector.contract.negotiation.ConsumerContractNegotiationManagerImpl.sendOffer(ConsumerContractNegotiationManagerImpl.java:279)
	at org.eclipse.dataspaceconnector.contract.negotiation.ConsumerContractNegotiationManagerImpl.processRequesting(ConsumerContractNegotiationManagerImpl.java:305)
	at org.eclipse.dataspaceconnector.spi.telemetry.Telemetry.lambda$contextPropagationMiddleware$0(Telemetry.java:73)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:258)
	at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:248)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.count(ReferencePipeline.java:709)
	at org.eclipse.dataspaceconnector.common.statemachine.StateProcessorImpl.process(StateProcessorImpl.java:49)
	at java.base/java.util.stream.ReferencePipeline$5$1.accept(ReferencePipeline.java:231)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.LongPipeline.reduce(LongPipeline.java:498)
	at java.base/java.util.stream.LongPipeline.sum(LongPipeline.java:456)
	at org.eclipse.dataspaceconnector.common.statemachine.StateMachine.performLogic(StateMachine.java:118)
	at org.eclipse.dataspaceconnector.common.statemachine.StateMachine.lambda$loop$2(StateMachine.java:106)
	at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:79)
	at io.micrometer.core.instrument.Timer.lambda$wrap$0(Timer.java:160)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

### Expected Behavior

- if there is a problem transforming the data, it should not be logged using log level `FINE`
- if there is a problem transforming the data, it should just be logged once and the negotiation should go into some kind of error state
- contract negotiation should work as expected, when the `policyId` field in the DataManagementApi field is used.

### Observed Behavior

- 

## Steps to Reproduce

Example payload to send to the data management API
```json
{
    "connectorId": "foo",
    "connectorAddress": "<connector-address>",
    "offer": {
        "offerId": "<contract-definition-id>:foo",
        "assetId": "<asset-id>",
        "policyId": "<policy-id>"
    }
}
```

## Context Information
-

## Detailed Description
-

## Possible Implementation
-

### LINKED ISSUE #1340: policyId is not mapped correctly from/to EDC/IDS
# Bug Report

## Describe the Bug
When the data catalog of a participant is requested, the `policyId` field of the `ContractOffer` does not match the actual one.

### Expected Behavior
To receive the correct policy id

### Observed Behavior
The `ContractOfferToIdsContractOfferTransformer` does not map the policyId to the IDS message and the `IdsContractOfferToContractOfferTransformer` creates a policy object without specifying the `id` field, doing so, a new random UUID is used.

## Steps to Reproduce
1. Start two connectors
2. Create an Asset, a Policy and a Contract Definition on one connector
3. request the catalog of the first connector from the second connector
4. verify that the `policyId` in the contract offer does not match the `policyId` of the Policy created at point `2.`

## Context Information
-

## Detailed Description
-

## Possible Implementation
-

### PR DESCRIPTION
**Pull request by paullatzelsperger**
## What this PR changes/adds

This PR removes the `uid` property from `Policy`, and therefore introduces another object, called `PolicyDefinition`, which contains that `uid` and a `Policy` property.

Policies as masterdata got replaced by that `PolicyDefinition` object.

Conversely, when policies are persisted as properties of `ContractNegotiation`s and `ContractAgreement`s, they do not need an ID. 

TL;DR:
- removes `Policy#uid`
- adds `PolicyDefinition`, which is a wrapper for `Policy`
- `ContractAgreement` and `ContractOffer` now have an embedded `Policy` property for persistence
- `PolicyArchive` does not obtain its `Policy` from the `PolicyStore` anymore, it merely returns the one attached to the `ContractAgreement`

## Why it does that

When consumer and provider are negotiating a contract, a new policy can be created in every ping-pong, which certainly would have to result in a new ID. 
Now, this would also mean that all those (or at least the final) policies need to be persisted in the `PolicyStore` on both sides, which would have undesirable effects, e.g. it would also expose them through the DM API.

## Further notes

- The composite ("wrapper") pattern has been chosen over inheritance to avoid problems with JSON serialization/deserialization: through polymorphism a `ContractOffer` could end up with a `PolicyDefinition`, which would cause Jackson to fail with an unrecognized property.
- the possibility to initialize a contract negotiation "by policyId" was removed 

## Linked Issue(s)

Closes #1340 
Closes #1251 

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [x] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ConsumerContractNegotiationManagerImpl.java [hunk 1]
```diff
@@ -125,10 +121,10 @@ public StatusResult<ContractNegotiation> initiate(ContractOfferRequest contractO
      * ContractNegotiation and transitions the ContractNegotiation to CONSUMER_APPROVING,
      * CONSUMER_OFFERING or DECLINING.
      *
-     * @param token Claim token of the consumer that send the contract request.
+     * @param token         Claim token of the consumer that send the contract request.
```
Comment (jimmarino) [inline core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ConsumerContractNegotiationManagerImpl.java:124]: formatting

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ConsumerContractNegotiationManagerImpl.java [hunk 2]
```diff
@@ -125,10 +121,10 @@ public StatusResult<ContractNegotiation> initiate(ContractOfferRequest contractO
...
      * @param negotiationId Id of the ContractNegotiation.
      * @param contractOffer The contract offer.
-     * @param hash A hash of all previous contract offers.
+     * @param hash          A hash of all previous contract offers.
```
Comment (jimmarino) [inline core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ConsumerContractNegotiationManagerImpl.java:127]: formatting

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ProviderContractNegotiationManagerImpl.java [hunk 3]
```diff
@@ -84,16 +84,11 @@ public void stop() {
...
-    public void enqueueCommand(ContractNegotiationCommand command) {
-        commandQueue.enqueue(command);
-    }
-
     /**
      * Tells this manager that a {@link ContractNegotiation} has been declined by the counter-party.
      * Transitions the corresponding ContractNegotiation to state DECLINED.
      *
-     * @param token Claim token of the consumer that sent the rejection.
+     * @param token         Claim token of the consumer that sent the rejection.
```
Comment (jimmarino) [inline core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ProviderContractNegotiationManagerImpl.java:91]: formatting

**Review by jimmarino — APPROVED** [review]
LGTM, just fix the minor formatting inconsistencies.

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/offer/ContractDefinitionServiceImpl.java [hunk 4]
```diff
@@ -91,7 +94,7 @@ private boolean evaluatePolicies(ContractDefinition definition, ParticipantAgent
     private Result<Policy> evaluate(String policyId, ParticipantAgent agent) {
         return Optional.of(policyId)
                 .map(policyStore::findById)
-                .map(policy -> policyEngine.evaluate(NEGOTIATION_SCOPE, policy, agent))
+                .map(policy -> policyEngine.evaluate(NEGOTIATION_SCOPE, policy.getPolicy(), agent))
```
Comment (ndr-brt) [inline core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/offer/ContractDefinitionServiceImpl.java:97]: `policy.getPolicy()` gets pretty weird to read, I would move this getter on a `.map` function, to have this written as `.map(PolicyDefinition::getPolicy)`

File: spi/policy-spi/src/main/java/org/eclipse/dataspaceconnector/spi/policy/store/PolicyStore.java [hunk 5]
```diff
@@ -15,6 +15,7 @@
 package org.eclipse.dataspaceconnector.spi.policy.store;
```
Comment (ndr-brt) [inline spi/policy-spi/src/main/java/org/eclipse/dataspaceconnector/spi/policy/store/PolicyStore.java:15]: Shouldn't this be called `PolicyDefinitionStore` for coherence?

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/PolicyApi.java [hunk 6]
```diff
@@ -18,19 +18,19 @@
...
 import jakarta.validation.Valid;
 import org.eclipse.dataspaceconnector.api.query.QuerySpecDto;
-import org.eclipse.dataspaceconnector.policy.model.Policy;
+import org.eclipse.dataspaceconnector.policy.model.PolicyDefinition;
 
 import java.util.List;
 
 @OpenAPIDefinition
 @Tag(name = "Policy")
 public interface PolicyApi {
```
Comment (ndr-brt) [inline extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/PolicyApi.java:27]: Will be more time consuming, but it would be better to add a policydefinition api module and deprecate the policy one, will be more easy for users to handle the breaking change

**Review by ndr-brt — COMMENTED** [review]
Pretty much good with this, I think the difference between `Policy` and `PolicyDefinition` should be more sharp, to avoid misunderstandings

File: spi/policy-spi/src/main/java/org/eclipse/dataspaceconnector/spi/policy/store/PolicyStore.java (hunk 5 above)
Comment (paullatzelsperger) [reply spi/policy-spi/src/main/java/org/eclipse/dataspaceconnector/spi/policy/store/PolicyStore.java:15]: yup, you're right. will refactor.

File: extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/PolicyApi.java (hunk 6 above)
Comment (paullatzelsperger) [reply extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/PolicyApi.java:27]: as far as I am aware, there was a general understanding that breaking changes are "allowed" without deprecation. In this case it would probably mean to maintain the "old" and new policy DTOs.

What we could do as next-best-thing is create a git tag right before merging this PR and publishing the artifacts, to give people a somewhat up-to-date version

Comment (ndr-brt) [reply extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/PolicyApi.java:27]: ok, my proposal would be to rename the api to `PolicyDefinitionApi`, the basepath to `policydefinitions` and the controller class accordingly. the module is already called `policydefinition` so.. effort earned :)

Comment (paullatzelsperger) [reply extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/PolicyApi.java:27]: I would like to avoid this unless there is a clear and present requirement from any of the projects to have the "old" policy API in the latest upstream `main`.

we're miles away from a "final" version, so I think the expectation from the general public to have stable APIs is not really justified.

Comment (ndr-brt) [reply extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/PolicyApi.java:27]: @paullatzelsperger forget about the "breaking change", this comment could be summarized as:
- `PolicyApi` class should be renamed to `PolicyDefinitionApi` and the related controller as well
- the api `@Path` should be moved from `/policies` to `/policydefinitions`

Comment (paullatzelsperger) [reply extensions/api/data-management/policydefinition/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/PolicyApi.java]: done. also renamed the extension to `PolicyDefinitionApiExtension` to keep consistency.

**Review by ndr-brt — APPROVED** [review]
